app/js/UrlCache.js

```diff
@@ -12,264 +12,54 @@
-                project_id,
-                url,
-                lastModified,
-                (error) => {
-                  if (error != null) {
-                    return callback(error)
-                  }
-                  return callback(
-                    null,
-                    UrlCache._cacheFilePathForUrl(project_id, url)
-                  )
-                }
-              )
-            }
-          )
-        } else {
-          logger.log({ url, lastModified }, 'URL is up to date in cache')
-          return callback(null, UrlCache._cacheFilePathForUrl(project_id, url))
-        }
-      }
-    )
-  },
-
-  _doesUrlNeedDownloading(project_id, url, lastModified, callback) {
-    if (callback == null) {
-      callback = function (error, needsDownloading) {}
-    }
-    if (lastModified == null) {
-      return callback(null, true)
-    }
-    return UrlCache._findUrlDetails(project_id, url, function (
-      error,
-      urlDetails
-    ) {
-      if (error != null) {
-        return callback(error)
-      }
-      if (
-        urlDetails == null ||
-        urlDetails.lastModified == null ||
-        urlDetails.lastModified.getTime() < lastModified.getTime()
-      ) {
-        return callback(null, true)
-      } else {
-        return callback(null, false)
-      }
-    })
-  },
-
-  _cacheFileNameForUrl(project_id, url) {
-    return project_id + ':' + crypto.createHash('md5').update(url).digest('hex')
-  },
-
-  _cacheFilePathForUrl(project_id, url) {
-    return `${Settings.path.clsiCacheDir}/${UrlCache._cacheFileNameForUrl(
-      project_id,
-      url
-    )}`
-  },
-
-  _clearUrlFromCache(project_id, url, callback) {
-    if (callback == null) {
-      callback = function (error) {}
-    }
-    return UrlCache._clearUrlDetails(project_id, url, function (error) {
-      if (error != null) {
-        return callback(error)
-      }
-      return UrlCache._deleteUrlCacheFromDisk(project_id, url, function (
-        error
-      ) {
-        if (error != null) {
-          return callback(error)
-        }
-        return callback(null)
-      })
-    })
-  },
+function getProjectDir(projectId) {
+  return Path.join(Settings.path.clsiCacheDir, projectId)
+}
 
-  _deleteUrlCacheFromDisk(project_id, url, callback) {
-    if (callback == null) {
-      callback = function (error) {}
-    }
-    return fs.unlink(UrlCache._cacheFilePathForUrl(project_id, url), function (
-      error
-    ) {
-      if (error != null && error.code !== 'ENOENT') {
-        // no error if the file isn't present
-        return callback(error)
-      } else {
-        return callback()
-      }
-    })
-  },
+function getCachePath(projectId, url, lastModified) {
+  // The url is a filestore URL.
+  // It is sufficient to look at the path and mtime for uniqueness.
+  const mtime = (lastModified && lastModified.getTime()) || 0
+  const key = new URL(url).pathname.replace(/\//g, '-') + '-' + mtime
```

Comment:
We don't have any restrictions on the pathname, so it could be arbitrarily long(*).  Maybe using the md5 hash as before is not such a bad idea?

(*)  (and contain arbitrary characters, which doesn't matter in principle as long we never run a shell command on them)

Reply:
With the host override to filestore in place, the URL pathname should be `/project/_MONGO_ID_/file/_other_mongo_id_` only.
https://github.com/overleaf/clsi/blob/af0b3eab6588c004c31fe8312657a4b1e277b540/app/js/UrlFetcher.js#L45-L48

WDYT about validating the URL with a simple regex `^/project/[a-f0-9]{24}/file/[a-f0-9]{24}$` instead of the _expensive_ hashing?

Reply:
>WDYT about validating the URL with a simple regex ^/project/[a-f0-9]{24}/file/[a-f0-9]{24}$ instead of the expensive hashing?

Yes that would be a good alternative.  The clsi API is meant to allow external urls, but I don't think we use that in production.  Maybe put the check in the RequestParser so we reject any unexpected urls up front?

Reply:
If we implement sending signed urls from web then we may need to adjust this.